broker/src/test/java/io/camunda/zeebe/broker/jobstream/JobGatewayStreamerTest.java

```diff
@@ -0,0 +1,198 @@
+
+  private final JobGatewayStreamer streamer =
+      new JobGatewayStreamer(eventService, communicationService, registry, random);
+
+  @RegisterExtension
+  private final ControlledActorSchedulerExtension scheduler =
+      new ControlledActorSchedulerExtension();
+
+  @BeforeEach
+  void beforeEach() {
+    scheduler.submitActor(streamer);
+    scheduler.workUntilDone();
+  }
+
+  @AfterEach
+  void afterEach() {
+    CloseHelper.quietCloseAll(
+        () -> {
+          streamer.closeAsync();
+          scheduler.workUntilDone();
+        });
+  }
+
+  @Test
+  void shouldNotifyAvailableJobsWhenNoConsumers() {
+    // given
+    final var type = BufferUtil.wrapString("foo");
+
+    // when
+    final var maybeStream = streamer.streamFor(type);
+    scheduler.workUntilDone();
+
+    // then
+    assertThat(maybeStream).isEmpty();
+    Mockito.verify(eventService, Mockito.times(1))
+        .broadcast(JobStreamTopics.JOB_AVAILABLE.topic(), "foo");
+  }
+
+  @Test
+  void shouldPickRandomTarget() {
+    // given - a registry which returns a set of consumers sorted by their member IDs
+    final var type = new UnsafeBuffer(BufferUtil.wrapString("foo"));
+    final var firstStreamId = new StreamId(UUID.randomUUID(), MemberId.from("a"));
+    final var secondStreamId = new StreamId(UUID.randomUUID(), MemberId.from("b"));
+    final var metadata = new DummyActivationProperties();
+    final SortedSet<StreamConsumer<JobActivationProperties>> consumers =
+        new TreeSet<>(Comparator.comparing(c -> c.id().receiver()));
+    consumers.add(new StreamConsumer<>(firstStreamId, metadata, type));
+    consumers.add(new StreamConsumer<>(secondStreamId, metadata, type));
+    registry.consumers.put(type, consumers);
+
+    // when - we fix that we always go for the second consumer, then push
+    random.fixedLong = 1L << 32;
+    final var stream = streamer.streamFor(type).orElseThrow();
+    stream.push(new TestJob(1, new JobRecord()), (job, error) -> {});
+    scheduler.workUntilDone();
+
+    // then
+    Mockito.verify(
+            communicationService,
+            Mockito.timeout(5_000).times(1).description("should pick the second stream"))
+        .send(
+            Mockito.eq(JobStreamTopics.PUSH.topic()),
+            Mockito.any(),
+            Mockito.any(),
+            Mockito.any(),
+            Mockito.eq(secondStreamId.receiver()),
+            Mockito.any());
+  }
+
+  @Test
+  void shouldReportCorrectMetadata() {
+    // given
+    final var type = new UnsafeBuffer(BufferUtil.wrapString("foo"));
+    final var streamId = new StreamId(UUID.randomUUID(), MemberId.from("a"));
+    final var metadata = new DummyActivationProperties();
+    registry.consumers.put(type, Set.of(new StreamConsumer<>(streamId, metadata, type)));
+
+    // when
+    final var stream = streamer.streamFor(type).orElseThrow();
+
+    // then
+    assertThat(stream.metadata()).isSameAs(metadata);
+  }
+
+  @Test
+  void shouldPush() {
```

Comment:
Not super happy with all the mocks, but we're still missing some parts for integration tests (i.e. the engine parts). Once we have those and we write some integration tests, we can look at the overlap and maybe remove these.